src/Lean/Elab/Tactic/ElabTerm.lean

```diff
@@ -381,36 +387,80 @@ private partial def blameDecideReductionFailure (inst : Expr) : MetaM Expr := wi
-      catch _ =>
-        diagnose expectedType s none
+    let r ← withAtLeastTransparency .default <| whnf s
+    if r.isAppOf ``isTrue then
+      -- Success!
+      -- While we have a proof from reduction, we do not embed it in the proof term,
+      -- and instead we let the kernel recompute it during type checking from the following more
+      -- efficient term. The kernel handles the unification `e =?= true` specially.
+      return pf
     else
-      let r ← withAtLeastTransparency .default <| whnf s
-      if r.isAppOf ``isTrue then
-        -- Success!
-        -- While we have a proof from reduction, we do not embed it in the proof term,
-        -- and instead we let the kernel recompute it during type checking from the following more
-        -- efficient term. The kernel handles the unification `e =?= true` specially.
-        return pf
-      else
-        diagnose expectedType s r
-where
+      diagnose expectedType s r
+  doKernel (expectedType : Expr) : TacticM Expr := do
+    let pf ← mkDecideProof expectedType
+    -- Get instance from `pf`
+    let s := pf.appFn!.appArg!
+    -- Reduce the decidable instance to (hopefully!) `isTrue` by passing `pf` to the kernel.
+    -- The `mkAuxLemma` function caches the result in two ways:
+    -- 1. First, the function makes use of a `type`-indexed cache per module.
+    -- 2. Second, once the proof is added to the environment, the kernel doesn't need to check the proof again.
+    let levelsInType := (collectLevelParams {} expectedType).params
+    -- Level variables occurring in `expectedType`, in ambient order
+    let lemmaLevels := (← Term.getLevelNames).reverse.filter levelsInType.contains
+    try
+      let lemmaName ← mkAuxLemma lemmaLevels expectedType pf
+      return mkConst lemmaName (lemmaLevels.map .param)
+    catch _ =>
+      diagnose expectedType s none
+  doNative (expectedType : Expr) : TacticM Expr := do
+    let d ← mkDecide expectedType
+    let levelsInType := (collectLevelParams {} expectedType).params
+    -- Level variables occurring in `expectedType`, in ambient order
+    let levels := (← Term.getLevelNames).reverse.filter levelsInType.contains
+    let auxDeclName ← Term.mkAuxName `_nativeDecide
+    let decl := Declaration.defnDecl {
+      name := auxDeclName
+      levelParams := levels
+      type := mkConst ``Bool
+      value := d
+      hints := .abbrev
+      safety := .safe
+    }
+    addAndCompile decl
+    -- get instance from `d`
+    let s := d.appArg!
+    let rflPrf ← mkEqRefl (toExpr true)
+    let pf := mkApp3 (mkConst ``of_decide_eq_true) expectedType s <| mkApp3 (mkConst ``Lean.ofReduceBool) (mkConst auxDeclName) (toExpr true) rflPrf
+    try
+      let lemmaName ← mkAuxLemma levels expectedType pf
+      return .const lemmaName (levels.map .param)
+    catch _ =>
+      throwError "\
+        tactic '{tacticName}' evaluated that the proposition\
+        {indentExpr expectedType}\n\
+        is false"
```

Comment:
Is that really the only reason that this can throw? I’m worried that the error message may be misleading if there are other reasons why this might fail.

Can we evaluate `auxDeclName` (using [Lean.evalConst](https://loogle.lean-lang.org/?q=Lean.evalConst)) to check?

Reply:
That's a good idea to evaluate the constant on error as a diagnostic check.